interfaces/connection.go

```diff
@@ -0,0 +1,255 @@
+}
+
+// Apps returns all the apps associated with this slot.
+func (slot *ConnectedSlot) Apps() map[string]*snap.AppInfo {
+	return slot.slotInfo.Apps
+}
+
+// Hooks returns all the hooks associated with this slot.
+func (slot *ConnectedSlot) Hooks() map[string]*snap.HookInfo {
+	return slot.slotInfo.Hooks
+}
+
+// SecurityTags returns the security tags for this slot.
+func (slot *ConnectedSlot) SecurityTags() []string {
+	return slot.slotInfo.SecurityTags()
+}
+
+// StaticAttr returns a static attribute with the given key, or error if attribute doesn't exist.
+func (slot *ConnectedSlot) StaticAttr(key string) (interface{}, error) {
+	if val, ok := slot.staticAttrs[key]; ok {
+		return val, nil
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
+// StaticAttrs returns all static attributes.
+func (slot *ConnectedSlot) StaticAttrs() map[string]interface{} {
+	return slot.staticAttrs
+}
+
+// Attr returns a dynamic attribute with the given name. It falls back to returning static
+// attribute if dynamic one doesn't exist. Error is returned if neither dynamic nor static
+// attribute exist.
+func (slot *ConnectedSlot) Attr(key string) (interface{}, error) {
+	if slot.dynamicAttrs != nil {
+		if val, ok := slot.dynamicAttrs[key]; ok {
+			return val, nil
+		}
+	}
+	return slot.StaticAttr(key)
+}
+
+// Attrs returns all dynamic attributes of this slot.
+func (slot *ConnectedSlot) Attrs() map[string]interface{} {
+	if slot.dynamicAttrs == nil {
+		slot.dynamicAttrs = make(map[string]interface{})
+	}
+	return slot.dynamicAttrs
+}
+
+// SetAttr sets the given dynamic attribute. Error is returned if the key is already used by a static attribute.
+func (slot *ConnectedSlot) SetAttr(key string, value interface{}) error {
+	if _, ok := slot.staticAttrs[key]; ok {
+		return fmt.Errorf("cannot change attribute %q as it was statically specified in the snap details", key)
+	}
+	if slot.dynamicAttrs == nil {
+		slot.dynamicAttrs = make(map[string]interface{})
+	}
+	// make a normalized copy
+	slot.dynamicAttrs[key] = copyRecursive(value)
+	return nil
+}
+
+// Interface returns the name of the interface for this connection.
+func (conn *Connection) Interface() string {
+	return conn.plug.plugInfo.Interface
+}
+
+func copyAttributes(value map[string]interface{}) map[string]interface{} {
+	return copyRecursive(value).(map[string]interface{})
+}
+
+func copyRecursive(value interface{}) interface{} {
+	// Copy all attributes and normalize ints/floats using their 64-bit variants.
+	// That kind of normalization happens in normalizeYamlValue(..) for static attributes
+	// when the yaml is loaded, but it needs to be done here as well because we're also
+	// dealing with dynamic attributes (and attributes set by the code of interfaces).
+	//
+	// note: ensure all the mutable types (or types that need a conversion)
+	// are handled here.
+	switch v := value.(type) {
+	case int:
+		return int64(v)
+	case float32:
+		return float64(v)
```

Comment:
My review is unhandled and unresponded to here.

Reply:
afaiu, now copyRecursive is used in SetAttr as well, so it helps maintaining  the invariant setup in snap/info_snap_yaml.go that attributes don't contain anything but int64 and float64 for new values being set. Without this kind of code in some form, interfaces themselves cannot continue to assume that to be the case.

Reply:
This code has been changed to make just a recursive copy of the static attirbutes, and normalize the dynamic attributes in the constructor and in SetAttr methods.